### PR TITLE
deprecation and vulnerability - add fields to search documents

### PR DESCRIPTION
**Pull request by camigthompson**
Addresses: https://github.com/NuGet/Engineering/issues/4501

### CODE REVIEW

File: src/NuGet.Jobs.Db2AzureSearch/NuGet.Jobs.Db2AzureSearch.csproj [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (drewgillies) [inline src/NuGet.Jobs.Db2AzureSearch/NuGet.Jobs.Db2AzureSearch.csproj:1]: This is odd. :)

File: src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs [hunk 2]
```diff
@@ -74,5 +74,6 @@ public class BaseMetadataDocument : CommittedDocument, IBaseMetadataDocument
         /// </summary>
         [SearchableField(AnalyzerName = PackageIdCustomAnalyzer.Name)]
         public string TokenizedPackageId { get; set; }
+
```
Comment (drewgillies) [inline src/NuGet.Services.AzureSearch/Models/BaseMetadataDocument.cs:77]: As is this. You can revert this file for cleanness.

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 3]
```diff
@@ -398,6 +400,8 @@ private V3SearchPackage ToV3SearchPackage(SearchDocument.Full result, string reg
                         };
                     })
                     .ToList(),
+                DeprecationInfo = result.DeprecationInfo.Message != null ? result.DeprecationInfo : null,
```
Comment (drewgillies) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:403]: Should `DeprecationInfo` be null checked?

Comment (drewgillies) [reply src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:403]: And `Vulnerabilities`?

**Review by drewgillies — DISMISSED** [review]
Some nits--the null checks are the critical pieces there but you may have covered this upstream.

**Comment by joelverhagen** [comment]
I think we should merge this into a feature branch. If this get's merged to `dev`, then taken to DEV, INT, then PROD, couldn't the existing service and jobs break since the C# models refer to Azure Search fields that don't exist in the running index? We can talk offline about the rollout/phasing strategy and use a similar approach as @advay26's search-by-TFM work.

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs [hunk 4]
```diff
@@ -146,4 +151,27 @@ public LatestFlags(LatestVersionInfo latest, bool isLatestStable, bool isLatest)
             public bool IsLatest { get; }
         }
     }
+
+    public class Deprecation
+    {
+        public AlternatePackage AlternatePkg { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/Models/SearchDocument.cs:157]: Does calling it `AlternatePackage` work? This abbreviation is not used in NuGet typically.

File: src/NuGet.Services.SearchService.Core/appsettings.json [hunk 5]
```diff
@@ -14,4 +14,4 @@
     }
   },
   "AllowedHosts": "*"
-}
+}
```
Comment (joelverhagen) [inline src/NuGet.Services.SearchService.Core/appsettings.json:17]: nit: revert unnecessary change

File: src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs [hunk 6]
```diff
@@ -357,5 +360,21 @@ private static void PopulateKey(KeyedDocument document, string packageId, Search
         {
             document.IsExcludedByDefault = isExcludedByDefault;
         }
+
+        private static void PopulateDeprecationInfo(
+            SearchDocument.Full document,
+            Package package)
+            {
+                document.DeprecationInfo = new Deprecation();
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs:368]: This is clearly placeholder code (which is fine for an incremental PR like this). Could you put a TODO linking to a GitHub issue that tracks the work that will put this in its final, expected state?

**Review by joelverhagen — CHANGES_REQUESTED** [review]
I left some small comments. Let's merge this to a feature branch, no `dev` since it's not production ready yet (more PRs will be needed).

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs (hunk 4 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/Models/SearchDocument.cs:157]: `AlternatePackageInfo` would also be consistent with `DeprecationInfo` if we want to separate the type from the variable name

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using Azure.Search.Documents.Indexes;
+using System.Collections.Generic;
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/Models/SearchDocument.cs:5]: We always place System.* libraries as first ones.

File: src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs [hunk 8]
```diff
@@ -357,5 +360,21 @@ private static void PopulateKey(KeyedDocument document, string packageId, Search
...
+                document.DeprecationInfo.AlternatePkg = new AlternatePackage();
+                document.DeprecationInfo.Reasons = Array.Empty<string>();
+            }
+
+        private static void PopulateVulnerabilities(
+            SearchDocument.Full document,
+            Package package)
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/SearchDocumentBuilder.cs:375]: >package [](http://example.com/codeflow?start=20&length=7)

Is package used so far?

File: src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchPackage.cs [hunk 9]
```diff
@@ -66,5 +67,11 @@ public class V3SearchPackage
         public List<V3SearchVersion> Versions { get; set; }
 
         public object Debug { get; set; }
+
+        [JsonPropertyName("deprecation")]
+        public Deprecation DeprecationInfo { get; set; }
+
+        [JsonPropertyName("vulnerabilities")]
+        public List<Vulnerability> Vulnerabilities { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchPackage.cs:75]: We should decouple the Azure Search data models from the response models. This allows us to be _conscious_ of API/protocol changes apart from internal implementation details (model changes). Said another way, once we ship this API, we really change change the response model without coordination with the client software. But we're free to change how the Azure Search index is shaped as long as we roll it out safely (index rebuilds, new code + new index type thing).

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs (hunk 4 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/Models/SearchDocument.cs]: I'm curious what the benefit is of having the "Info" suffix on the name. Wouldn't `AlternatePackage` and `Deprecation` work for the property names?

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 10]
```diff
@@ -398,6 +400,8 @@ private V3SearchPackage ToV3SearchPackage(SearchDocument.Full result, string reg
                         };
                     })
                     .ToList(),
+                DeprecationInfo = (result.DeprecationInfo != null && result.DeprecationInfo.Message != null) ? result.DeprecationInfo : null,
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:403]: Consider adding a private help method to do the Deprecation model mapping and another to do the vulnerabilities (or per vulnerability instance) mapping.

File: src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchPackage.cs (hunk 9 above)
Comment (camigthompson) [reply src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchPackage.cs:75]: So essentially have separate Deprecation/Vulnerability types for V3SearchPackage and SearchDocument?

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchPackage.cs:75]: That's right. There will be some minor differences, like `[JsonPropertyName(...)]` on the V3SearchPackage related models.

File: src/NuGet.Services.AzureSearch/Models/SearchDocument.cs (hunk 4 above)
Comment (camigthompson) [reply src/NuGet.Services.AzureSearch/Models/SearchDocument.cs]: I was thinking that "Info" suffix just keeps the type and the variable name separate, but if it's unnecessary then I can make that change.

File: src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchDeprecation.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text.Json.Serialization;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class V3SearchDeprecation
+    {
+
+        [JsonPropertyName("alternatePackage")]
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchDeprecation.cs:11]: nit: let's remove the unnecessary lines here.

File: src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchVulnerability.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text.Json.Serialization;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class V3SearchVulnerability
+    {
+
+        [JsonPropertyName("advisoryUrl")]
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/SearchService/Models/V3SearchVulnerability.cs:11]: nit: let's remove the unnecessary lines in this file.

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 13]
```diff
@@ -437,6 +441,46 @@ private string GetIconUrl(IBaseMetadataDocument document)
...
         }
 
+        private V3SearchDeprecation GetV3SearchDeprecation(SearchDocument.Full document)
+        {
+            if (document.Deprecation != null)
+            {
+                var alternatepackage = new V3SearchAlternatePackage{};
+                var deprecation = new V3SearchDeprecation();  
+
+                alternatepackage.Id = document.Deprecation.AlternatePackage.Id;
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:451]: >.Id [](http://example.com/codeflow?start=75&length=3)

Is that possible that "AlternatePackage" is null here? If so, it will throw the exception. If "AlternatePackage" can be null as expected, the protection is needed here.

Comment (agr) [reply src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:451]: And it should be covered by tests.

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 14]
```diff
@@ -437,6 +441,46 @@ private string GetIconUrl(IBaseMetadataDocument document)
...
+            {
+                foreach (var vulnerability in document.Vulnerabilities)
+                {
+                    vulnerabilities.Add(new V3SearchVulnerability 
+                    {
+                        AdvisoryURL = vulnerability.AdvisoryURL,
+                        Severity = vulnerability.Severity
+                    }
+                    );
+                }
```
Comment (drewgillies) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:478]: super-nit: Generally we go with `});` in these cases.

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 15]
```diff
@@ -437,6 +441,46 @@ private string GetIconUrl(IBaseMetadataDocument document)
...
+            if (document.Vulnerabilities != null)
+            {
+                foreach (var vulnerability in document.Vulnerabilities)
+                {
+                    vulnerabilities.Add(new V3SearchVulnerability 
+                    {
+                        AdvisoryURL = vulnerability.AdvisoryURL,
+                        Severity = vulnerability.Severity
+                    }
+                    );
```
Comment (drewgillies) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:477]: super-nit: Generally we go with `});` in these cases.

File: src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs [hunk 16]
```diff
@@ -437,6 +441,52 @@ private string GetIconUrl(IBaseMetadataDocument document)
             return document.IconUrl;
         }
 
+        private V3SearchDeprecation GetV3SearchDeprecation(SearchDocument.Full document)
+        {
+            if (document.Deprecation != null)
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/SearchService/SearchResponseBuilder.cs:446]: Consider adding unit tests to cover all these corner cases (Deprecation + Vulnerabilities). This is to ensure that commits in the future will not break what we have today or at least get a warning before changing it.

**Review by joelverhagen — APPROVED** [review]
Add UTs for the `if` conditions added to response building (as @zhhyu mentioned) then :shipit:

**Review by joelverhagen — APPROVED** [review]

**Review by advay26 — APPROVED** [review]
